add-branch/Micah

micah-wehrle opened the pull request:
I am not sure if I'm doing this right. I'll try to copy all Tyler's crap.

Featured in this PR:
- Added universal-tabbed-card component and corresponding child universal-tab component.
- Added sample html to app.component.html to demonstrate the tabbed card.

Files touched:
- app.module.ts (added two new components)
- app.component.html (added sample html for new components)
- added contents of components/universal-tabbed-card/

I should note that the tabbed card is not styled how I want it (I'm not terrific at styling). My goal was to get it functioning as I wanted, and I have done that. I tried to make it as idiot-proof as possible and done truthy checks where needed (too often?)

Here's some sample images:
This is what I was trying to replicate from the original site:
<img width="402" alt="Screen Shot 2022-09-04 at 3 21 04 PM" src="https://user-images.githubusercontent.com/11700686/188506108-8c25ca22-a0ab-45db-9d8e-86217a38a67d.png">

This is what I created:
![image](https://user-images.githubusercontent.com/11700686/188506195-0a0bf735-1b6b-40e8-8e9c-780ccab09345.png)

How this tabbed card is implemented:
![Messages Image(1799361669)](https://user-images.githubusercontent.com/11700686/188506738-61e61fe5-130f-4e36-a777-fe66920f8d66.png)
`</app-universal-tabbed-card>`

Here is the build info? Prod something? Sept 5 at 3:45pm
<img width="686" alt="Screen Shot 2022-09-05 at 3 42 26 PM" src="https://user-images.githubusercontent.com/11700686/188506386-68519ed7-ebd7-4b8d-a334-d08d83a57d23.png">

Please let me know anything I did wrong. Thanks

micah-wehrle committed:
Added tabbed card and tab components

micah-wehrle commented:
I know you wouldn't put test or sample material into a pr. We can cut it if you fancy.

looksliketyler commented on src/app/app.component.html:
```diff
@@ -1,3 +1,31 @@
 <app-navbar></app-navbar>
 <app-jumbotron></app-jumbotron>
 <router-outlet></router-outlet>
+<br>
```
lines 1-3 are all the html tags that will reside here - the below code should be inside its own component. i think this stuff was in services offered on old site

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tab/universal-tab.component.ts:
```diff
@@ -0,0 +1,16 @@
+})
+export class UniversalTabComponent {
+
+  @Input() tabTitle: string = 'Tab';
```
this is hardcoded to `'Tab'`, why?

StreamSG replied:
FTFY

micah-wehrle replied:
Not sure if it'd be better to leave it blank, but in case something goes wrong, this would be the default value. So the tab would be called 'Tab' instead of just nothing.

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tab/universal-tab.component.html:
```diff
@@ -0,0 +1,12 @@
+    role="tabpanel"
+    [attr.aria-labelledby]="'pills-'+ tabNum + '-' + parentUniqueId +'-tab'"
+    tabindex="0"
+  >
```
this can be on line 8

micah-wehrle replied:
If this is standard (or Atlas) html formatting then fiiiine, but it's helpful for me to have a matching "closing" marker. In my mind, putting it on line 8 would be like:
`if(value === true) {`
&emsp;`&nbsp;console.log('it true');`
&emsp;`&nbsp;value = false; }`

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tab/universal-tab.component.ts:
```diff
@@ -0,0 +1,16 @@
+  styleUrls: ['./universal-tab.component.scss']
+})
+export class UniversalTabComponent {
+
```
remove this line

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tab/universal-tab.component.ts:
```diff
@@ -0,0 +1,16 @@
+export class UniversalTabComponent {
+
+  @Input() tabTitle: string = 'Tab';
+  @Input() active: boolean = false;
```
`active` is for changing the css class of active right?  there may be a way to change the approach to work with that rather than creating a global variable - if you keep it, try to be a little more descriptive with the variable name

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.html:
```diff
@@ -0,0 +1,20 @@
+<div class="container card">
+  <ul class="nav nav-pills" id="pills-tab" role="tablist">
+    <li class="nav-item" role="presentation" *ngFor="let tab of childTabs; let i = index">
```
`*ngFor` should only be inside `<ng-container>`

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.html:
```diff
@@ -0,0 +1,20 @@
+        role="tab"
+        [attr.aria-controls]="'pills-'+ i +'-'+ uniqueId"
+        [attr.aria-selected]="(i === activeTab ? 'active' : '')"
+      >{{ tabTitles[i] }}</button>
```
seems like youre using Prettier extension?  yucky

micah-wehrle replied:
I don’t know. Are you talking about attr.? Or the aria-stuff? The aria stuff is part of bootstrap I think, I got it from the sample tab setups on the bootstrap site. As far as attr., I found that I needed it to property bind to those attributes. If you do, for example, aria-selected=“” then you can’t put any logic in the “” (like with property binding). But if you do [aria-selected] it will say property doesn’t exist for that element (also tried ariaSelected). I found online that you just need to do [attr.aria-selected] to be able to property bind.

Again, just what I found by looking online. As far as I could tell that was the only way to use the bootstrap styling with the angular property binding. If we can find a better way I’m down, I’m just ignorant about it.

micah-wehrle replied:
Ok now knowing what prettier is, no sorry I just put them all on their own lines. Would you have formatted it differently? I won't use Prettier

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.html:
```diff
@@ -0,0 +1,20 @@
+    </li>
+  </ul>
+  <div class="tab-content">
+    <ng-content></ng-content>
```
youll have to learn me about `<ng-content>`  it may be used in the udemy, but im not super familiar since we dont use it in atlas

StreamSG replied:
ng-content is basically a placeholder for HTML that allows you to throw HTML into the component depending on what's needed. If you check the services-offered-component it calls <app-universal-tabbed-card> and then adds tabs and HTML between the opening and closing tags.

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+import { AfterContentInit, Component, ContentChildren, QueryList } from '@angular/core';
+import { UniversalTabComponent } from './universal-tab/universal-tab.component';
```
should have a blank space for line 2 - want to separate package imports and local imports

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+  styleUrls: ['./universal-tabbed-card.component.scss']
+})
+export class UniversalTabbedCardComponent implements AfterContentInit {
+
```
remove line 10

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+  // Get all the child tabs within this component
+  @ContentChildren(UniversalTabComponent) childTabs: QueryList<UniversalTabComponent> = new QueryList<any>();
+  
+  tabTitles: string[] = [];
```
missing access modifer `public` or `private` - same for line 15

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+
+  constructor() {}
+
+  ngAfterContentInit(): void {
```
hm...why use `ngAfterContentInit()`?

micah-wehrle replied:
Because I'm using ContentChild/ren to grab the universal-tab components nested within this component, you have to use ngAfterContentInit instead of ngOnInit because they ElementRefs for the universal-tab components won't yet be initialized. Although I am grabbing them to set a few variables within the component classes (and correct for a few potential user errors), I need to wait until after the content is initialized before I can do so. Hence ngAfterContentInit. ContentChild/ren can't be used with ngOnInit or you'll get undefined values.

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+          childTab.tabNum = i; // Inform each tab of its numerical order 
+          childTab.parentUniqueId = this.uniqueId; // Inform each tab of this tabbed-card's unique id to prevent two tabbed cards from interacting with each other
+          if(childTab.active) {
+            if(!foundActiveTab) {
```
in atlas, we dont nest more than 3 logics - so this line here is nest 4

micah-wehrle replied:
My disappointment is immeasurable and my day is ruined. Doing the following for `this.childTabs` and `childTab` within the forEach be possible/reasonable?

`if(typeof this.childTabs === 'undefined') return;`
`if(typeof childTab === 'undefined') return;`

looksliketyler commented on src/app/components/universal-tabbed-card/universal-tabbed-card.component.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+  }
+
+}
```
it seems theres a lot of logic here for some stuff you can let bootstrap css/js handle - but awesome job coding it out

StreamSG committed:
Services offered fix

Replaced code from app component to proper component

StreamSG committed:
Couple of fixes

Adjusted component for universal tabbed card, adjusted ngFor to be used within ng-container, cleaned up code

micah-wehrle commented on src/app/views/services-offered/services-offered.component.html:
```diff
@@ -1 +1,25 @@
-<p>services-offered works!</p>
+<p>Below is just to demonstrate how the tabbed card works. Still needs style work.</p>
```
I appreciate you doing this as Tyler asked.

I would push back though against the idea of putting just this into it's own component. The design for a universal component, in my understanding, is that the content of the component is determined as it's used. So the user may want multiple tabbed-cards within the application. Would they then make a specific component as a sort of wrapper for each one, and then just have the content for that specific card like how you have it here?

I think you could definitely argue yes, it really depends on the application. I personally wouldn't do it this way. I would just find whatever component represents that specific page, and then put all this html in in there. 

Note: I wouldn't just shove this all into the app.component (as it was originally), that was just for demonstration.

looksliketyler commented on src/app/app.component.html:
```diff
@@ -1,3 +1,3 @@
 <app-navbar></app-navbar>
 <app-jumbotron></app-jumbotron>
-<router-outlet></router-outlet>
```
WHY WAS THIS REMOVED?!

micah-wehrle replied:
👀 
@StreamSG was this you
